Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,387 @@
+  mkEmpty 5
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+
+/-- Create vector `#[v₁, v₂, v₃, v₄, v₅, v₆]` -/
+def mkVector6 {α : Type u} (v₁ v₂ v₃ v₄ v₅ v₆ : α) : Vector α 6 :=
+  mkEmpty 6
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+
+/-- Create vector `#[v₁, v₂, v₃, v₄, v₅, v₆, v₇]` -/
+def mkVector7 {α : Type u} (v₁ v₂ v₃ v₄ v₅ v₆ v₇ : α) : Vector α 7 :=
+  mkEmpty 7
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+    |> push v₇
+
+/-- Create vector `#[v₁, v₂, v₃, v₄, v₅, v₆, v₇, v₈]` -/
+def mkVector8 {α : Type u} (v₁ v₂ v₃ v₄ v₅ v₆ v₇ v₈ : α) : Vector α 8 :=
+  mkEmpty 8
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+    |> push v₇
+    |> push v₈
+
+/-- `extract v start halt` Returns the slice of `v` from indices `start` to `stop` (exclusive).
+  If `start` is greater or equal to `stop`, the result is empty.
+  If `stop` is greater than the size of `v`, the size is used instead.-/
+def extract {α : Type u_1} (v : Vector α n)
+  (start : Nat) (halt : Nat) : Vector α (min halt n - start) :=
+  ⟨Array.extract v.val start halt, proof⟩
+  where
+    proof := by
+      simp only [size_extract]
+      have size_eq : n = v.val.size := by
+        exact v.property.symm
+      rw [←size_eq]
+      done
+
+
+/--Eliminator for `Vector α n`-/
+@[elab_as_elim]
+def elim {α} {C : ∀ {n}, Vector α n → Sort u}
+    (H : ∀ a : Array α, C ⟨a, rfl⟩) {n : Nat} : ∀ v : Vector α n, C v
+  | ⟨a, h⟩ =>
+    match n, h with
+    | _, rfl => H a
+
+
+/-- Map a vector under a function. -/
+def map (f : α → β) : Vector α n → Vector β n
+  | ⟨a, h⟩ => ⟨Array.map f a, by simp [*]⟩
+
+
+/-- Mapping two vectors under a curried function of two variables. -/
+def map₂ (f : α → β → φ) : Vector α n → Vector β n → Vector φ n
+  | ⟨a, h₁⟩, ⟨b, h₂⟩ => ⟨Array.zipWith a b f, by simp [Array.size_zipWith, h₁, h₂]⟩
+
+
+/- OVERLAP : `mkVector`-/
+/-- Vector obtained by repeating an element -/
+def replicate (n : Nat) (a : α) : Vector α n := mkVector n a
+
+/-- Drop `i` elements from a vector of length `n`; we can have `i > n`. -/
+def drop (i : Nat) : Vector α n → Vector α (n - i)
+  | ⟨{data := l}, p⟩ => ⟨{data := List.drop i l}, by simp [*]⟩
+
+/-- Take `i` elements from a vector of length `n`; we can have `i > n`. -/
+def take (i : Nat) : Vector α n → Vector α (min i n)
+  | ⟨{data := l}, p⟩ => ⟨{data := List.take i l}, by simp [*]⟩
+
+
+
+/-- Vector of length `n` from a function on `Fin n`. -/
+def ofFn : ∀ {n}, (Fin n → α) → Vector α n
+  | 0, _ => nil
+  | _ + 1, f => cons (f 0) (ofFn fun i ↦ f i.succ)
+
+/-- Create a vector from another with a provably equal length. -/
+protected def congr {n m : Nat} (h : n = m) : Vector α n → Vector α m
+  | ⟨x, p⟩ => ⟨x, h ▸ p⟩
+
+
+-- Swap functions
+/--
+Swaps two entries in an array, or panics if either index is out of bounds.
+
```

Comment:
Is there a reason why 

```lean
def swap (v : Vector α n) (i j : Fin n) : Vector α n :=
  ⟨Array.swap v.val ⟨i.val, sorry⟩ ⟨j.val, sorry⟩, sorry⟩
```

is not also defined?

Reply:
I should add it. Thanks :) . This section comes from the basic array defs file. The idea being to achieve the maximum possible feature parity with Arrays. I must missed it when borrowing function signatures from there.

Reply:
I have fixed this in the next commit.